File: 21mergeTwoSortedLists/mergeTwoSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        std::forward_list<decltype(std::declval<ListNode>().val)> linkedList;
        auto position{linkedList.cbefore_begin()};
        while (l1 && l2)
        {
            if (l1->val < l2->val)
            {
                position = linkedList.emplace_after(position, l1->val);
                l1 = l1 ->next;
            }
            else
            {
                position = linkedList.emplace_after(position, l2 -> val);
                l2 = l2 ->next;
            }
        }
        auto l{l1 ? l1 : l2};
        while (l)
        {
            position = linkedList.emplace_after(position, l->val);
            l = l->next;
        }
        auto const result{std::make_unique<ListNode>(0)};
        auto current{result.get()};
        std::for_each(std::cbegin(linkedList), std::cend(linkedList), [&current](auto const _){
            current->next = new ListNode{_};
            current = current->next;});
        return result->next;
    }
};

File: 21mergeTwoSortedLists/mergeTwoSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        final LinkedList<Integer> linkedList = new LinkedList<>();
        while (l1 != null && l2 != null)
        {
            if (l1.val < l2.val)
            {
                linkedList.offer(l1.val);
                l1 = l1.next;
            }
            else
            {
                linkedList.offer(l2.val);
                l2 = l2.next;
            }
        }
        ListNode l = l1 != null ? l1 : l2;
        while (l != null)
        {
            linkedList.offer(l.val);
            l = l.next;
        }
        final ListNode result = new ListNode(0);
        ListNode current = result;
        for (final int $: linkedList)
        {
            current.next = new ListNode($);
            current = current.next;
        }
        return result.next;
    }
}
